google-cloud-spanner/src/test/java/com/google/cloud/spanner/DatabaseClientImplTest.java

```diff
@@ -195,6 +198,177 @@ public void tearDown() {
     mockSpanner.removeAllExecutionTimes();
   }
 
+  @Test
+  public void
+      testPoolMaintainer_whenInactiveTransactionAndSessionIsNotFoundOnBackend_removeSessionsFromPool() {
+    InactiveTransactionRemovalOptions inactiveTransactionRemovalOptions =
+        InactiveTransactionRemovalOptions.newBuilder()
+            .setIdleTimeThreshold(
+                Duration.ofSeconds(2L)) // any session idle for more than 2s will be long-running
+            .setExecutionFrequency(Duration.ofSeconds(1L)) // check long-running sessions every 1s
+            .build();
+    SessionPoolOptions sessionPoolOptions =
+        SessionPoolOptions.newBuilder()
+            .setMinSessions(1)
+            .setMaxSessions(1) // to ensure there is 1 session and pool is 100% utilized
+            .setCloseIfInactiveTransactions()
+            .setInactiveTransactionRemovalOptions(inactiveTransactionRemovalOptions)
+            .setLoopFrequency(1000L) // main thread runs every 1s
+            .build();
+    spanner =
+        SpannerOptions.newBuilder()
+            .setProjectId(TEST_PROJECT)
+            .setDatabaseRole(TEST_DATABASE_ROLE)
+            .setChannelProvider(channelProvider)
+            .setCredentials(NoCredentials.getInstance())
+            .setSessionPoolOption(sessionPoolOptions)
+            .build()
+            .getService();
+    DatabaseClientImpl client =
+        (DatabaseClientImpl)
+            spanner.getDatabaseClient(DatabaseId.of(TEST_PROJECT, TEST_INSTANCE, TEST_DATABASE));
+    Instant initialExecutionTime = client.pool.poolMaintainer.lastExecutionTime;
+
+    try (TransactionManager manager = client.transactionManager()) {
+      TransactionContext transaction = manager.begin();
+      mockSpanner.setOnCommitThrowSessionNotFoundException(true);
+      while (true) {
+        try {
+          // Simulate a delay of 4s to ensure that the below transaction is a long-running one.
+          // We require to wait for 4s so that the main thread executes at-least once every 1s
+          // As per this test, anything which takes more than 2s is long-running
+          mockSpanner.setExecuteSqlExecutionTime(
+              SimulatedExecutionTime.ofMinimumAndRandomTime(
+                  (int) Duration.ofSeconds(4).toMillis(), 0));
+          transaction.executeUpdate(UPDATE_STATEMENT);
```

Comment:
Could we try to solve this using the following strategy instead:
1. Add `Clock` as a package-private option to `SessionPoolOptions`, so we can set that in a test like this as well.
2. Then instead of adding a long wait to an RPC, just manually advance the clock after executing the statement, and then force a maintenance loop. Note: The long-running transaction handler is in real-life not very likely to kick in while an RPC is running, it is much more likely to kick in while there is nothing more going on, because the application has leaked the session.

Regarding point 2 above: Ideally, the solution that we are adding here should not even kick in at all in this scenario, as the transaction is actually still running and has an active request. The fact that this request takes a 'long' time is something that should not be seen as inactivity. Instead, if the statement runs for too long, it is the deadline for that specific statement that should let it fail. After that, the transaction could be eligible for being regarded as inactive.

Reply:
Hmmm.... Coming back on my comment on point 2 above: It is probably not really a problem if we allow sessions with an RPC to be cleaned up by this task. Although it's not the optimal solution, it's probably the easiest, and as far as I can see it won't cause any problems, because:
1. The fact that the session is removed from the pool while still active does not make the session unusable for the transaction that is already using it. It only means that it won't be used by any other transactions in the future.
2. We're apparently also able to handle the fact that the transaction tries to return a session to the pool that the pool no longer knows.

Reply:
Yes, I have added this strategy now. And have removed the need to introduce delays in mock RPC call.